kube-core/src/labels.rs

```diff
@@ -0,0 +1,649 @@
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Equal("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo=bar")
+    /// ```
+    Equal(String, String),
+
+    /// Key does not exists or is not equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotEqual("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo!=bar")
+    /// ```
+    NotEqual(String, String),
+
+    /// Key exists:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Exists("foo".into()).to_string();
+    /// assert_eq!(exp, "foo")
+    /// ```
+    Exists(String),
+
+    /// Key does not exist:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::DoesNotExist("foo".into()).to_string();
+    /// assert_eq!(exp, "!foo")
+    /// ```
+    DoesNotExist(String),
+}
+
+/// Perform selection on a list of expressions
+#[derive(Clone, Debug, Eq, PartialEq, Default, Deserialize, Serialize)]
+pub struct Selector(Expressions);
+
+impl Selector {
+    /// Create a selector from a vector of expressions
+    fn from_expressions(exprs: Expressions) -> Self {
+        Self(exprs)
+    }
+
+    /// Create a selector from a map of key=value label matches
+    fn from_map(map: BTreeMap<String, String>) -> Self {
+        Self(map.into_iter().map(|(k, v)| Expression::Equal(k, v)).collect())
+    }
+
+    /// Indicates whether this label selector matches all pods
+    pub fn selects_all(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Extend the list of expressions for the selector
+    ///
+    /// ```
+    /// use kube_core::Selector;
+    /// use kube_core::Expression;
+    /// use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+    ///
+    /// let label_selector: Selector = LabelSelector::default().try_into().unwrap();
+    /// let mut selector = &mut Selector::default();
+    /// selector = selector.extend(Expression::Equal("environment".into(), "production".into()));
+    /// selector.extend([Expression::Exists("bar".into()), Expression::Exists("foo".into())].into_iter());
+    /// selector.extend(label_selector);
+    /// ```
+    pub fn extend(&mut self, exprs: impl IntoIterator<Item = Expression>) -> &mut Self {
+        self.0.extend(exprs);
+        self
+    }
+}
+
+impl Matcher for LabelSelector {
+    /// Perform best effort match on label selector
+    ///
+    /// ```
+    /// use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+    /// use kube_core::Matcher;
+    ///
+    /// LabelSelector::default().matches(&Default::default());
+    /// ```
+    fn matches(&self, labels: &BTreeMap<String, String>) -> bool {
+        let selector: Result<Selector, ParseExpressionError> = self.clone().try_into();
+        match selector {
+            Ok(selector) => selector.matches(labels),
+            Err(_) => false,
+        }
+    }
```

Comment:
If this is best effort, maybe it's better to encourage it being turned into one of our structs first (with a `?`) rather than discarding the error internally

we should be able to `labelselector.try_into()?` and do matches on that right?

Reply:
Added some docstring comments with such scenarios. A bit wacky because of the `&impl Matcher` as it can’t infer type, but is working.